sdks/python/apache_beam/examples/inference/anomaly_detection/anomaly_detection_pipeline/pipeline/transformations.py

```diff
@@ -0,0 +1,194 @@
+class ModelWrapper(DistilBertModel):
+  """Wrapper to DistilBertModel to get embeddings when calling
+    forward function."""
+  def forward(self, **kwargs):
+    output = super().forward(**kwargs)
+    sentence_embedding = (
+        self.mean_pooling(output,
+                          kwargs["attention_mask"]).detach().cpu().numpy())
+    return sentence_embedding
+
+  # Mean Pooling - Take attention mask into account for correct averaging
+  def mean_pooling(self, model_output, attention_mask):
+    """
+        Calculates the mean of token embeddings
+
+        Args:
+          model_output: The output of the model.
+          attention_mask: This is a tensor that contains 1s for all input tokens and
+          0s for all padding tokens.
+
+        Returns:
+          The mean of the token embeddings.
+        """
+    token_embeddings = model_output[
+        0]  # First element of model_output contains all token embeddings
+    input_mask_expanded = (
+        attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float())
+    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(
+        input_mask_expanded.sum(1), min=1e-9)
+
+
+# [END DistilBertModelWrapper]
+
+
+# [START CustomSklearnModelHandlerNumpy]
+class CustomSklearnModelHandlerNumpy(SklearnModelHandlerNumpy):
+  # Can be removed once: https://github.com/apache/beam/issues/21863 is fixed
+  def batch_elements_kwargs(self):
+    """Limit batch size to 1 for inference"""
+    return {"max_batch_size": 1}
+
+  # Can be removed once: https://github.com/apache/beam/issues/22572 is fixed
+  def run_inference(self, batch, model, inference_args=None):
+    """Runs inferences on a batch of numpy arrays.
+
+        Args:
+          batch: A sequence of examples as numpy arrays. They should
+            be single examples.
+          model: A numpy model or pipeline. Must implement predict(X).
+            Where the parameter X is a numpy array.
+          inference_args: Any additional arguments for an inference.
+
+        Returns:
+          An Iterable of type PredictionResult.
+        """
+    _validate_inference_args(inference_args)
+    vectorized_batch = np.vstack(batch)
+    predictions = hdbscan.approximate_predict(model, vectorized_batch)
+    return [PredictionResult(x, y) for x, y in zip(batch, predictions)]
+
+
+# [END CustomSklearnModelHandlerNumpy]
+
+
+class NormalizeEmbedding(beam.DoFn):
+  """A DoFn for normalization of text embedding."""
+  def process(self, element, *args, **kwargs):
+    """
+        For each element in the input PCollection, normalize the embedding vector, and
+        yield a new element with the normalized embedding added
+
+        Args:
+          element: The element to be processed.
+        """
+    (text, uid), prediction = element
+    embedding = prediction.inference
+    l2_norm = np.linalg.norm(embedding)
+    yield {"text": text, "id": uid, "embedding": embedding / l2_norm}
+
+
+class Decode(beam.DoFn):
```

Comment:
Needs more description , below we have DecodePrediction, maybe here this can be DecodeExample or DecodeInferenceData...